File: TranslationFiestaRuby/lib/translation_fiesta/infrastructure/app_paths.rb
# frozen_string_literal: true

require 'fileutils'

module TranslationFiesta
  module Infrastructure
    module AppPaths
      module_function

      def app_root
        @app_root ||= begin
          executable_dir = File.dirname(File.expand_path($PROGRAM_NAME))
          find_app_root_from(executable_dir) || File.expand_path('../../../..', __dir__)
        end
      end

      def data_root
        @data_root ||= begin
          override = ENV['TF_APP_HOME']
          path = override.nil? || override.strip.empty? ? File.join(app_root, 'data') : override
          ensure_dir(path)
        end
      end

      def logs_dir
        ensure_dir(File.join(data_root, 'logs'))
      end

      def exports_dir
        ensure_dir(File.join(data_root, 'exports'))
      end

      def settings_path
        File.join(data_root, 'settings.json')
      end

      def memory_db_path
        File.join(data_root, 'translation_memory.db')
      end

      def ensure_dir(path)
        FileUtils.mkdir_p(path)
        path
      end

      def find_app_root_from(start_dir)
        current = start_dir
        loop do
          return current if app_root_candidate?(current)

          parent = File.expand_path('..', current)
          return nil if parent == current

          current = parent
        end
      end
      private_class_method :find_app_root_from

      def app_root_candidate?(directory)
        File.exist?(File.join(directory, 'bin', 'translation_fiesta')) &&
          File.directory?(File.join(directory, 'lib', 'translation_fiesta'))
      end
      private_class_method :app_root_candidate?

      private_class_method :ensure_dir
    end
  end
end

File: TranslationFiestaRust/src/app_paths.rs
use std::env;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};

#[derive(Debug, Clone)]
pub struct AppPaths {
    pub app_root: PathBuf,
    pub data_root: PathBuf,
    pub logs_dir: PathBuf,
    pub exports_dir: PathBuf,
    pub settings_file: PathBuf,
    pub memory_db_file: PathBuf,
}

impl AppPaths {
    pub fn discover() -> Result<Self> {
        let app_root = resolve_app_root()?;
        let data_root = resolve_data_root(&app_root)?;
        let logs_dir = ensure_dir(data_root.join("logs"))?;
        let exports_dir = ensure_dir(data_root.join("exports"))?;
        let settings_file = data_root.join("settings.json");
        let memory_db_file = data_root.join("translation_memory.db");

        Ok(Self {
            app_root,
            data_root,
            logs_dir,
            exports_dir,
            settings_file,
            memory_db_file,
        })
    }
}

fn resolve_app_root() -> Result<PathBuf> {
    if let Ok(exe) = env::current_exe()
        && let Some(parent) = exe.parent()
    {
        return Ok(parent.to_path_buf());
    }

    env::current_dir().context("failed to resolve current directory")
}

fn resolve_data_root(app_root: &Path) -> Result<PathBuf> {
    if let Ok(override_path) = env::var("TF_APP_HOME") {
        let trimmed = override_path.trim();
        if !trimmed.is_empty() {
            return ensure_dir(PathBuf::from(trimmed));
        }
    }

    ensure_dir(app_root.join("data"))
}

fn ensure_dir(path: PathBuf) -> Result<PathBuf> {
    std::fs::create_dir_all(&path)
        .with_context(|| format!("failed to create directory {}", path.display()))?;
    Ok(path)
}
